mautic/mautic
# Additional filter operators for Segments and Reports

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Related user documentation PR URL | Pending
| Related developer documentation PR URL | Pending
| Issues addressed (#s or URLs) | Pending
| BC breaks? | N/A
| Deprecations? | N/A

### To Do


[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

The `regexp` and `notRegexp` filters are difficult for non-devs to use, as the syntax is rather complicated. This PR adds some commonly used options (`contains`, `startsWith`, and `endsWith`) as operators for both Segment and Report text-based filters. There usage is as follows:

Remember that with SQL, `%` is the wildcard character.

##### contains
This will function nearly identical to the `like` operator. The main difference is that with `contains`, the value entered in the text box will ALWAYS be wrapped in `%`. With `like`, it was only wrapped with `%` if the field value did not already start with `%`. So you can have a filter of `last name contains "lbe"` to pull in all contacts that match `%lbe%`.

##### startsWith
This function will append `%` to the value entered in the text box. So you can have a filter of `last name starts with "Gil"` to pull in all contacts that match `Gil%`.

##### endsWith
This function will prepend `%` to the value entered in the text box. So you can have a filter of `last name ends with "bert"` to pull in all contacts that match `%bert`.

As a safety feature, this PR fixes a potential issue with `generateRandomParameterName()` to ensure that we never generate the same random parameter name as one already used. It's very much a corner case, but the code added to that method eliminates the possibility.

This PR also fixes the long standing issues where dynamic report filters would cause the set report filters to not be used at all, and instead filter only on any of the set dynamic filters.

[//]: # ( As applicable: )
#### Steps to test this PR:
1. Have contacts in your database.
2. Create a new segment with a text-based filter, and use `starts with` as the operator. Enter a value in the text box that you would expect to match contacts in your system.
3. Update the segment via `app/console m:s:u -i X`, where `X` equals the ID of the segment you created.
4. You should see contacts added. Verify they are correct in the UI
5. Repeat steps 2-4 using text-based filters with the `ends with` operator.
6. Repeat steps 2-4 using text-based filters with the `contains` operator.
7. Repeat steps 5 & 6 using dynamic filters from the dropdown tab at the top of the report view/details page. Ensure that the query being built is the one expected and that the data returned is as expected as well.

#### List deprecations along with the new alternative:
1. It is as yet undecided whether or not we will deprecate the `like` functionality. Probably not though.
